src/archive/src/main.rs

```diff
@@ -0,0 +1,438 @@
+    last_upgrade_timestamp: Timestamp,
+}
+
+impl Storable for ConfigState {
+    fn to_bytes(&self) -> Cow<[u8]> {
+        match &self {
+            ConfigState::Uninitialized => Cow::Borrowed(&[]),
+            ConfigState::Initialized(config) => {
+                let buf = candid::encode_one(config).expect("failed to encode archive config");
+                Cow::Owned(buf)
+            }
+        }
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        if bytes.len() == 0 {
+            return ConfigState::Uninitialized;
+        }
+        ConfigState::Initialized(
+            candid::decode_one::<ArchiveConfig>(&bytes).expect("failed to decode archive config"),
+        )
+    }
+}
+
+/// Index key for the anchor index.
+/// Changing the (serialized) size of this value requires a stable memory migration.
+#[derive(Debug)]
+struct AnchorIndexKey {
+    anchor: Anchor,
+    timestamp: Timestamp,
+    log_index: u64,
+}
+
+impl Storable for AnchorIndexKey {
+    fn to_bytes(&self) -> Cow<[u8]> {
+        let mut buf = Vec::with_capacity(std::mem::size_of::<AnchorIndexKey>());
+        buf.extend(&self.anchor.to_le_bytes());
+        buf.extend(&self.timestamp.to_le_bytes());
+        buf.extend(&self.log_index.to_le_bytes());
+        Cow::Owned(buf)
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        AnchorIndexKey {
+            anchor: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[0..8]).expect("failed to read anchor"),
+            ),
+            timestamp: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[8..16]).expect("failed to read timestamp"),
+            ),
+            log_index: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[16..]).expect("failed to read log_index"),
+            ),
+        }
+    }
+}
+
+#[update]
+fn write_entry(anchor: Anchor, timestamp: Timestamp, entry: ByteBuf) {
+    with_config(|config| {
+        if config.ii_canister != caller() {
+            trap(&format!(
+                "Only {} is allowed to write entries.",
+                config.ii_canister
+            ))
+        }
+    });
+    let idx = with_log(|log| {
+        log.append(entry.as_ref())
+            .expect("failed to append log entry")
+    });
+
+    with_anchor_index_mut(|index| {
+        let key = AnchorIndexKey {
+            anchor,
+            timestamp,
+            log_index: idx as u64,
+        };
+
+        // the only way this expect can trigger is when the key size is wrong.
+        // On other failures (e.g. no more stable memory available) the underlying StableBTreeMap will panic directly.
+        index.insert(key, ()).expect("bug: key size mismatch");
+    })
+}
+
+#[query]
+fn get_entries(index: Option<u64>, limit: Option<u16>) -> Entries {
+    let num_entries = limit_or_default(limit);
+
+    with_log(|log| {
+        let length = log.len();
+        let start_idx = match index {
+            None => length.saturating_sub(num_entries),
```

Comment:
If `index` is not specified, we only return the _last_ `limit` entries, is that correct? Why not return `0 .. limit`? Also would be good to mention this in the API description:

> If no limit is given, at most the configured number (see ArchiveInit) entries are returned.